File: TinkerRocketApp/TinkerRocketApp/Models/SensorCalStatus.swift
//
//  SensorCalStatus.swift
//  TinkerRocketApp
//
//  Decoded form of the FC's SensorCalStatusData frame (issue #132): the
//  gyro zero-rate bias + high-g accel bias currently stored in the rocket's
//  NVS.  Arrives on the file_ops characteristic behind a 0xCB discriminator
//  (sibling of the 0xCA mag-cal frame).  Wire layout mirrors
//  SensorCalStatusData in RocketComputerTypes.h — keep both in sync.
//
//  Wire (19 bytes, little-endian, packed):
//    [0]      uint8  valid       (1 if a cal is stored)
//    [1..2]   int16  gyro_x      (raw gyro LSB)
//    [3..4]   int16  gyro_y
//    [5..6]   int16  gyro_z
//    [7..10]  float  hg_x        (high-g accel bias, m/s²)
//    [11..14] float  hg_y
//    [15..18] float  hg_z
//

import Foundation

struct SensorCalStatus: Equatable {
    let valid: Bool
    let gyroX: Int16
    let gyroY: Int16
    let gyroZ: Int16
    let hgX: Float
    let hgY: Float
    let hgZ: Float

    /// Decode the payload *after* the 0xCB discriminator byte.
    static func decode(_ bytes: [UInt8]) -> SensorCalStatus? {
        guard bytes.count >= 19 else { return nil }
        func i16(_ o: Int) -> Int16 {
            bytes.withUnsafeBufferPointer {
                UnsafeRawBufferPointer($0).loadUnaligned(fromByteOffset: o, as: Int16.self)
            }
        }
        func f32(_ o: Int) -> Float {
            bytes.withUnsafeBufferPointer {
                UnsafeRawBufferPointer($0).loadUnaligned(fromByteOffset: o, as: Float.self)
            }
        }
        return SensorCalStatus(
            valid: bytes[0] != 0,
            gyroX: i16(1), gyroY: i16(3), gyroZ: i16(5),
            hgX: f32(7), hgY: f32(11), hgZ: f32(15))
    }
}

File: TinkerRocketAndroid/core/protocol/src/main/kotlin/com/tinkerbug/tinkerrocket/protocol/SensorCalStatus.kt
package com.tinkerbug.tinkerrocket.protocol

/**
 * SensorCalStatus — decoded form of the FC's SensorCalStatusData frame
 * (issue #132): the gyro zero-rate bias + high-g accel bias currently stored
 * in the rocket's NVS.  Arrives on the file_ops characteristic behind a 0xCB
 * discriminator (sibling of the 0xCA mag-cal frame).  Wire layout mirrors
 * SensorCalStatusData in RocketComputerTypes.h.
 *
 * Wire (19 bytes, little-endian, packed):
 *   [0]      u8   valid       (1 if a cal is stored)
 *   [1..2]   i16  gyro_x      (raw gyro LSB)
 *   [3..4]   i16  gyro_y
 *   [5..6]   i16  gyro_z
 *   [7..10]  f32  hg_x        (high-g accel bias, m/s²)
 *   [11..14] f32  hg_y
 *   [15..18] f32  hg_z
 */
public data class SensorCalStatus(
    val valid: Boolean,
    val gyroX: Int,       // i16 raw gyro LSB
    val gyroY: Int,
    val gyroZ: Int,
    val hgX: Float,       // high-g accel bias, m/s²
    val hgY: Float,
    val hgZ: Float,
) {
    public companion object {
        public const val SIZE: Int = 19

        /** Decode the payload *after* the 0xCB discriminator byte; null if too short. */
        public fun decode(payload: ByteArray): SensorCalStatus? {
            if (payload.size < SIZE) return null
            val b = LeBuffer(payload)
            return SensorCalStatus(
                valid = b.u8() != 0,
                gyroX = b.i16(), gyroY = b.i16(), gyroZ = b.i16(),
                hgX = b.f32(), hgY = b.f32(), hgZ = b.f32(),
            )
        }
    }
}
